Autoupdater improvements

NeoRaider opened the pull request:
The main change is the addition of the download.d directory, which allows to stop services after the manifest has been verified, but before the image is downloaded.

Suggested list of services to stop (RFC):
* uhttpd
* alfred
* respondd
* batman-adv-visdata
* haveged

jplitza commented on admin/autoupdater/files/usr/lib/lua/autoupdater/util.lua:
```diff
@@ -1,13 +1,28 @@
+module('autoupdater.util', package.seeall)
+
+
+-- Executes a command in the background, without parsing the command though a shell (in contrast to os.execute)
```
Typo: s/though/through/

jplitza commented on admin/autoupdater/files/usr/lib/lua/autoupdater/util.lua:
```diff
@@ -1,13 +1,28 @@
+function exec(...)
+  local pid, errno, error = nixio.fork()
+  if pid == 0 then
+    nixio.execp(...)
```
According to the [doc](http://luci.subsignal.org/api/nixio/modules/nixio.html#nixio.execp), `execp()` also launches a shell. Why not just `exec()`?

NeoRaider replied:
The docs are plain wrong. The difference between exec and execp is that exec expects an absolute path, while execp respects PATH (just like the execv and execvp C functions internally used by this).

jplitza commented on admin/autoupdater/files/usr/lib/lua/autoupdater/util.lua:
```diff
@@ -21,7 +36,41 @@ function popen(command)
     inw:close()
 
-    nixio.exec('/bin/sh', '-c', command)
+    nixio.execp(...)
```
See above.

jplitza commented on admin/autoupdater/files/usr/lib/lua/autoupdater/util.lua:
```diff
@@ -21,7 +36,41 @@ function popen(command)
+    end
+
+    local file = dir .. '/' .. entry
+    if fs.stat(file, 'type') ~= 'reg' then
```
This forbids symlinks. Intended?

NeoRaider replied:
fs.stat gives the information about the file linked to, to get the information about the link itself, fs.lstat must be used.

jplitza commented on admin/autoupdater/files/usr/lib/lua/autoupdater/util.lua:
```diff
@@ -21,7 +36,41 @@ function popen(command)
+    return true
+  end
+
+  local files = util.consume(fs.dir(dir))
```
Shouldn't this come after the following if-block?

NeoRaider replied:
Ah, the following was supposed to be `if not files`, will fix.

jplitza commented on admin/autoupdater/files/usr/sbin/autoupdater:
```diff
@@ -257,21 +259,25 @@ local function autoupdate(mirror)
 
-  os.execute('sync; sysctl -w vm.drop_caches=3')
+  autoupdater_util.exec('sync')
+  autoupdater_util.exec('sysctl', '-w', 'vm.drop_caches=3')
```
Why aren't these two commands a script in `download.d`?

NeoRaider replied:
Good idea.

jplitza commented on admin/autoupdater/files/usr/sbin/autoupdater:
```diff
@@ -257,21 +259,25 @@ local function autoupdate(mirror)
   end
 
-  local popen = io.popen(string.format("sha512sum '%s'", image))
+  local popen = io.popen(string.format("exec sha512sum '%s'", image))
```
Unrelated to the change at hand, but shouldn't we escape `'` in `image`? Or "simply" reimplement `io.popen()` the way we already have our `autoupdater_utils.popen()`, as passing the command as separate arguments would also solve the problem.

In general, I dislike that now we have some commands that are called as separate arguments and some that are still called via a shell as a single string.

NeoRaider replied:
True, reimplementing io.popen would be the best solution, but seemed like unnecessary work (as the Lua autoupdater will be replaced by C anyways after 2016.1).

The autoupdater_util.exec function, on the other hand, was important, as it is now also used to call the download.d and abort.d executables, which are supplied by other packages and thus might not follow sensible naming conventions ;)

jplitza commented on admin/autoupdater/files/usr/sbin/autoupdater:
```diff
@@ -257,21 +259,25 @@ local function autoupdate(mirror)
   if checksum ~= manifest.checksum then
     io.stderr:write('Invalid image checksum!\n')
     os.remove(image)
+    autoupdater_util.run_dir('/usr/lib/autoupdater/abort.d')
```
We should probably call `abort.d` around line 300 as well?

NeoRaider replied:
Indeed. I think making the path configurable is overkill, but having them defined at the top makes sense.

jplitza commented:
Done with the current version of code, waiting for fixes or responses.

Regarding the list of services to stop, this is what `sysupgrade` itself stops anyway on my node:
```
Sending TERM to remaining processes ... netifd crond gluon-crond uradvd uhttpd dnsmasq fastd fastd ntpd alfred batadv-vis dnsmasq respondd ubusd askfirst dhcpv6.script dhcpv6.script dhcpv6.script logd rpcd haveged 
```
Assuming we need to keep everything network-related (netifd, dnsmasq, fastd, dnsmasq, ubusd?, dhcp*), we could stop (in addition to your list):
* crond
* gluon-crond
* uradvd
* ntpd
* ubusd?
* askfirst
* logd?
* rpcd

jplitza commented:
Also, could you please simply commit the whitespace-fix? It only blows up this PR unnecessarily.

NeoRaider commented:
I've pushed the whitespace fix.

NeoRaider commented:
Updated branch pushed.

I'd like download.d not to unnecessarily cripple the node. As the download can take very long when the connection to the server is bad, I've kept my list to the monitoring services, so the node will disappear from the map, but apart from that continue to function.

* crond and gluon-crond might be stopped as well, I'm not sure about these. Would have the additional benefit to avoid starting multiple autoupdater instances
* No strong opinion about uradvd or ntpd...
* ubusd and logd are very central and can't be properly brought up in abort.d again once stopped
* askfirst is started by inittab, not an initscript, and is respawned as soon as it is terminated (which is why sysupgrade sends it a SIGKILL...)
* rpcd should not even be installed by default in the Gluon master, we've reverted the LuCI change adding a dependency on rpcd.

jplitza commented:
Sorry, my master version is a tad older, now that you mention it I remember that change removing rpcd again.

uradvd and ntpd don't really cripple the node: As it seems to have internet/mesh access (it successfully downloaded the manifest) chances are high another router with radvd running is available too. And ntpd isn't even visible to the outside world.

I understand your reservations about ubusd and logd, I had them too and thus marked them with a question mark.

askfirst probably doesn't really matter, but should be safe to somehow kill if we can stop init from restarting it.
